spring-cloud-starter-stream-sink-task-launcher-dataflow/README.adoc

```diff
@@ -0,0 +1,120 @@
+
+This application launches a registered task application using the Data Flow Server http://docs.spring.io/spring-cloud-dataflow/docs/current/reference/htmlsingle/#api-guide-resources-task-executions-launching[REST API].
+
+== Input
+
+Launch request parameters including:
+
+* the task name (required and registered as a task with the target Data Flow Server)
+* deployment properties (key value pairs, optional).
+* program arguments for the task (a list, optional).
+
+=== Headers:
+
+* `Content-Type: application/json`
+
+=== Payload:
+
+A JSON document:
+
+[source,json]
+----
+{
+  "name":"foo",
+  "deploymentProps": {"key1":"val1","key2":"val2"},
+  "args":["--debug", "--foo", "bar"]
+}
+----
+
+minimally,
+
+[source,json]
+----
+{"name":"foo"}
+----
+
+== Output
+
+N/A (launches task to the local system).
+
+=== Options
+
+The **$$tasklauncher-dataflow$$** $$sink$$ supports the following configuration properties:
+
+//tag::configuration-properties[]
+$$spring.cloud.dataflow.client.authentication.basic.password$$:: $$The login password.$$ *($$String$$, default: `$$<none>$$`)*
+$$spring.cloud.dataflow.client.authentication.basic.username$$:: $$The login username.$$ *($$String$$, default: `$$<none>$$`)*
+$$spring.cloud.dataflow.client.enable-dsl$$:: $$Enable Data Flow DSL access.$$ *($$Boolean$$, default: `$$false$$`)*
+$$spring.cloud.dataflow.client.server-uri$$:: $$The Data Flow server URI.$$ *($$String$$, default: `$$http://localhost:9393$$`)*
+$$spring.cloud.dataflow.client.skip-ssl-validation$$:: $$Skip Ssl validation.$$ *($$Boolean$$, default: `$$true$$`)*
+$$trigger.fixed-delay$$:: $$The fixed delay (polling interval).$$ *($$Integer$$, default: `$$1$$`)*
+$$trigger.initial-delay$$:: $$The initial delay.$$ *($$Integer$$, default: `$$1$$`)*
+$$trigger.time-unit$$:: $$The TimeUnit to apply to delay values.$$ *($$TimeUnit$$, default: `$$<none>$$`, possible values: `NANOSECONDS`,`MICROSECONDS`,`MILLISECONDS`,`SECONDS`,`MINUTES`,`HOURS`,`DAYS`)*
+//end::configuration-properties[]
+
+== Using the TaskLauncher
+A tasklauncher is a sink that consumes a `LaunchRequest` message, as described above, and launches the task on a
+target Spring Cloud Data Flow server (given by `--dataflow.uri`). The task launcher sink polls its input for launch
+requests but will pause when the SCDF server's task execution limit, `spring.cloud.dataflow.task
+.maximum-concurrent-tasks`
+is reached (see the
+http://docs.spring.io/spring-cloud-dataflow/docs/current/reference/htmlsingle/#spring-cloud-dataflow-task-limit-concurrent-executions[reference docs] for more details).
+
+When the number of running tasks drops below this limit, the sink will resume processing. This is intended to prevent
+the SCDF deployment platform from running out of resources under heavy task demand. The polling sink is also
+scheduled with a `PeriodicTrigger`. By default the maximum launch request rate is 1 per second, but may be
+configured to any duration, down to zero.
```

Comment:
I can see we want to launch at a relatively short interval when there are slots available but do we really want to pound the `serverIsAcceptingNewTasks()` every second when there are no slots?

Consider using a `DynamicPeriodicTrigger` and back-off the interval when there are no slots.

Reply:
"down to zero" - I wouldn't recommend that because we'd spin a CPU when there are no slots.

Reply:
I imagine there could be some scenarios where they would not want a delay, e.g., 1 sec is too long to wait.